### PR TITLE
WIP: Added Github workflow for frontend and backend

### PR DESCRIPTION
**Pull request by Minus20Five**
GitHub Issue (If applicable): #12

## PR Type

What kind of change does this PR introduce?
- Build or CI related changes
- Project automation

## What is the current behavior?
No CI 😢

## What is the new behavior?

Github workflow configured for both frontend and backend. We upload the `target/` directory containing the executable JAR as an artifact. See screenshots below.

Screen Shot 2020-03-08 at 8 56 04 PM
Screen Shot 2020-03-08 at 8 55 35 PM

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current upstream repo
- [ ] ~~Added new tests to the test suite for the feature~~ (N/A)
- [x] Tested the full test suite, and ensured that the feature does not break current build.
- [ ] Docs have been added/updated which fit (for bug fixes / features)
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable): N/A

### CODE REVIEW

**Comment by Matteas-Eden** [comment]
> 
> 
> Seeing the commit history (2 commits), it appears the commits have not been squashed into 1 commit. The assignment requires us to do this before a merge, unfortunately:
> "Prior to merging, all commits should be squashed. The reviewer should request the
> code author to do this if required. "

We can squash as part of the merge. GitHub handles that for us. That's also what the SQUASH label is for - to indicate whether a PR should be merged normally or with a squash. Come to think of it, maybe we should a rebase label as well...

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            ${{ runner.os }}-yarn-
+      - name: Yarn Install
+        working-directory: ./client
+        run: yarn install --frozen-lockfile
+      - name: Yarn Build
+        working-directory: ./client
+        run: yarn build
+      - name: Yarn Test
+        working-directory: ./client
+        run: CI=true yarn test --passWithNoTests
```
Comment (jordansimsmith) [inline .github/workflows/workflow.yml]: Should we test before we build? No point creating build files if the tests fail (build is creating production bundles)

Also should we upload the build files somewhere? Or is that the plan for the future. At the moment theres not really a use of this step.

Comment (Minus20Five) [reply .github/workflows/workflow.yml]: Yes for both.
Thanks for letter me know.

**Comment by Minus20Five** [comment]
Result of running [CI workflow](https://github.com/Minus20Five/softeng-701-assignment-1/runs/493210458?check_suite_focus=true) on my personal master branch.

File: .github/workflows/workflow.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        name: feeder-server
+        path: './server/target/'
+ 
+  client: # React CI
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Get Yarn Cache
+        id: yarn-cache-dir-path
```
Comment (PreetPatel) [inline .github/workflows/workflow.yml]: Do we want to specify a node version in the CI workflow? 
Something like
`- uses: actions/setup-node@v1`
`        with:`
`          node-version: "10.x"`
`          registry-url: "https://registry.npmjs.org"`

File: .github/workflows/workflow.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+      - name: Cache node_modules for Yarn
```
Comment (PreetPatel) [inline .github/workflows/workflow.yml]: Do we want to add code linting checks?
`- run: yarn lint`
Helps with code consistent code 
(If we define a linting rule)

File: .github/workflows/workflow.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      - name: Get Yarn Cache
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+      - name: Cache node_modules for Yarn 
+        uses: actions/cache@v1
+        id: yarn-cache
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/client/yarn.lock') }}
+          restore-keys: |
```
Comment (PreetPatel) [inline .github/workflows/workflow.yml]: Add code quality checkers? ie. SonarCloud etc

File: .github/workflows/workflow.yml (hunk 2 above)
Comment (PreetPatel) [reply .github/workflows/workflow.yml]: Imo this would be useful hence haven't approved

File: .github/workflows/workflow.yml (hunk 3 above)
Comment (Minus20Five) [reply .github/workflows/workflow.yml]: Sure. We can modify this step when we add a linter. That'll probably be done in a PR for issue #13

File: .github/workflows/workflow.yml (hunk 4 above)
Comment (Minus20Five) [reply .github/workflows/workflow.yml]: I'm not familiar with SonarCloud (looks like static analysis tool?). I think we should have a separate job to run static analysis + style checker (if we decide we need it). That can be done in a separate PR #13

File: .github/workflows/workflow.yml (hunk 2 above)
Comment (Minus20Five) [reply .github/workflows/workflow.yml]: Does React depend on Node.js?

Comment (PreetPatel) [reply .github/workflows/workflow.yml]: erp my bad

**Pull request merged**